theano/tensor/nnet/abstract_conv.py

```diff
@@ -320,13 +339,215 @@ def conv2d_grad_wrt_weights(input,
+        for dim in [0, 1, 2, 3]:
+            assert isinstance(input_shape[dim], (theano.tensor.TensorConstant,
+                                                 int, type(None)))
+
+    # setting the last two dimensions of filter_shape to None, if
+    # the type of these dimensions is TensorVariable.
+    numerical_filter_shape = list(filter_shape)
+    for dim in [2, 3]:
+        if isinstance(filter_shape[dim], theano.tensor.TensorVariable):
+            numerical_filter_shape[dim] = None
+
     gradWeight_op = AbstractConv2d_gradWeights(imshp=input_shape,
-                                               kshp=filter_shape,
+                                               kshp=numerical_filter_shape,
                                                border_mode=border_mode,
                                                subsample=subsample,
                                                filter_flip=filter_flip)
 
-    return gradWeight_op(input, output_grad, filter_shape)
+    return gradWeight_op(input, output_grad, filter_shape[:-2])
+
+
+def bilinear_kernel_2D(ratio, normalize=True):
+    """Compute 2D kernel for bilinear upsampling
+
+    This function builds the 2D kernel that can be used to upsample
+    a tensor by the given ratio using bilinear interpolation.
+
+    Parameters
+    ----------
+    ratio: int or Constant/Scalar Theano tensor the ratio by which an image
+        will be upsampled by the returned filter in the 2D space.
+
+    normalize: bool
+        param normalize: indicates whether to normalize the kernel or not.
+        Default is True.
+
+    Returns
+    -------
+    symbolic 2D tensor
+        the 2D kernels that can be applied to any given image to upsample it
+        by the indicated ratio using bilinear interpolation in two dimensions.
+
+    """
+
+    hkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle('x', 0)
+    vkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle(0, 'x')
+    kern = hkern * vkern
+    return kern
+
+
+def bilinear_kernel_1D(ratio, normalize=True):
+    """Compute 1D kernel for bilinear upsampling
+
+    This function builds the 1D kernel that can be used to upsample
+    a tensor by the given ratio using bilinear interpolation.
+
+    Parameters
+    ----------
+    ratio: int or Constant/Scalar Theano tensor the ratio by which an image
+        will be upsampled by the returned filter in the 2D space.
+
+    normalize: bool
+        param normalize: indicates whether to normalize the kernel or not.
+        Default is True.
+
+    Returns
+    -------
+    symbolic 1D tensor
+        the 1D kernels that can be applied to any given image to upsample it
+        by the indicated ratio using bilinear interpolation in one dimension.
+
+    """
+
+    T = theano.tensor
+    half_kern = T.arange(1, ratio + 1, dtype=theano.config.floatX)
+    kern = T.concatenate([half_kern, half_kern[-2::-1]])
+
+    if normalize:
+        kern /= float(ratio)
```

Comment:
This will not work when ratio is a tensor. Since `kern` is a floatX we don't need it though, so please remove the float cast. We should also test `bilinear_kernel_1D` for tensor ratio.